Signal/src/ViewControllers/MediaGalleryViewController.swift

```diff
@@ -168,43 +168,10 @@ protocol MediaGalleryDataSourceDelegate: class {
     func mediaGalleryDataSource(_ mediaGalleryDataSource: MediaGalleryDataSource, deletedSections: IndexSet, deletedItems: [IndexPath])
 }
 
-class MediaGalleryViewController: OWSNavigationController, MediaGalleryDataSource, MediaTileViewControllerDelegate {
-
-    private var pageViewController: MediaPageViewController?
-
-    private let uiDatabaseConnection: YapDatabaseConnection
-    private let editingDatabaseConnection: YapDatabaseConnection
-    private let mediaGalleryFinder: OWSMediaGalleryFinder
-
-    private var initialDetailItem: MediaGalleryItem?
-    private let thread: TSThread
-    private let options: MediaGalleryOption
-
-    // we start with a small range size for quick loading.
-    private let fetchRangeSize: UInt = 10
-
-    deinit {
-        Logger.debug("deinit")
-    }
-
-    @objc
-    init(thread: TSThread, uiDatabaseConnection: YapDatabaseConnection, options: MediaGalleryOption = []) {
-        self.thread = thread
-        assert(uiDatabaseConnection.isInLongLivedReadTransaction())
-        self.uiDatabaseConnection = uiDatabaseConnection
-
-        self.editingDatabaseConnection = OWSPrimaryStorage.shared().newDatabaseConnection()
-
-        self.options = options
-        self.mediaGalleryFinder = OWSMediaGalleryFinder(thread: thread)
-
-        super.init(nibName: nil, bundle: nil)
-    }
-
-    required init?(coder aDecoder: NSCoder) {
-        notImplemented()
-    }
+class MediaGalleryNavigationController: OWSNavigationController {
 
+    var presentationView: UIImageView!
+    var retainUntilDismissed: MediaGallery?
```

Comment:
I don't see this being cleared anywhere.  Will this by a retain cycle?  Are these entities being released?

Reply:
Yes, they're being released.

Ownership is basically inverted.

The MediaGallery has a *weak* reference to the MediaGalleryNavigationController it creates, which in turn has a *strong* reference back to the MediaGallery that created it.

So in effect, the lifetime of the media gallery is only as long as something *else* holds a reference to the MediaGalleryNavigationController. 

An alternative would be that everywhere we show the MediaGalleryNavigationController we also hold a reference to the MediaGallery, and then whenever we dismiss the MediaGalleryNavigationController, we also manually clear the reference to the MediaGallery.

While more explicit, that approach seemed more error prone.

I'm not crazy about the current approach, but I'm open to suggestion.

Reply:
If you prefer, I could also just revert c5bb924, and merge 97e2c33 which also works.